internal/backend/workloadproxy/reconciler.go

```diff
@@ -6,225 +6,374 @@
+		zap.Strings("upstreams", rd.GetHosts()),
+		zap.Stringer("probe_port", p),
+	)
+
+	got := rec.activeProbes.Get(clusterID)
+	if got == nil {
+		l, listErr := upstream.NewListWithCmp(asRemotes(h, p, listLogger), (*remote).Equal)
+		if listErr != nil {
+			return fmt.Errorf("failed to create upstream list for cluster %q: %w", clusterID, listErr)
+		}
 
-	if aliasToLB := registry.clusterToAliasToLBStatus[cluster]; aliasToLB == nil {
-		registry.clusterToAliasToLBStatus[cluster] = map[string]*lbStatus{}
+		got = rec.activeProbes.Set(clusterID, l, ttlcache.DefaultTTL)
 	}
 
-	registry.clusterToAliasToLBStatus[cluster][alias] = lbSts
+	got.Value().Reconcile(asRemotes(h, p, listLogger))
 
 	return nil
 }
 
 // GetProxy returns a proxy for the exposed service, targeting the load balancer for the given alias.
-func (registry *Reconciler) GetProxy(alias string) (http.Handler, resource.ID, error) {
-	registry.mu.Lock()
-	defer registry.mu.Unlock()
+func (rec *Reconciler) GetProxy(als string) (http.Handler, resource.ID, error) {
+	rec.mu.Lock()
+	defer rec.mu.Unlock()
+
+	al := alias(als)
+
+	c, p, ok := rec.aliasToCluster.ClusterPort(al)
+
+	logger := rec.logger.With(zap.String("cluster", c), zap.Stringer("alias", al), zap.Stringer("port", p))
+
+	logger.Debug("got proxy")
 
-	clusterID, ok := registry.aliasToCluster[alias]
 	if !ok {
 		return nil, "", nil
 	}
 
-	lbSts := registry.clusterToAliasToLBStatus[clusterID][alias]
-	if lbSts == nil || lbSts.lb == nil {
-		return nil, "", nil
+	proxy := httputil.NewSingleHostReverseProxy(&url.URL{Scheme: "http", Host: net.JoinHostPort(c, p.String())})
+	proxy.Transport = rec.transport
+	proxy.ErrorHandler = func(w http.ResponseWriter, req *http.Request, err error) {
+		logger.Error(
+			"proxy error",
+			zap.Error(err),
+			zap.String("cluster", c),
+			zap.Stringer("alias", al),
+			zap.String("alias_host", req.Host),
+		)
+
+		http.Error(w, "workload proxy error", http.StatusBadGateway)
 	}
 
-	hostPort := registry.hostPortForAlias(clusterID, alias)
+	return proxy, c, nil
+}
 
-	targetURL := &url.URL{
-		Scheme: "http",
-		Host:   hostPort,
+// All returns all the info about the load balancers. More info in the [Description] struct.
+func (rec *Reconciler) All() iter.Seq[Description] {
+	return func(yield func(Description) bool) {
+		rec.mu.Lock()
+		defer rec.mu.Unlock()
+
+		for alsPort, data := range rec.aliasToCluster.All() {
+			if !yield(Description{
+				Alias:     alsPort.F1.String(),
+				ClusterID: data.clusterID,
+				Port:      alsPort.F2.String(),
+				InUsePort: data.inUsePort.String(),
+				Upstream:  data.hosts,
+			}) {
+				return
+			}
+		}
 	}
+}
+
+// Description descibes an alias, its cluster, port and attached upstreams. It also shows if the current port is used
+// active probing.
+type Description struct {
+	Alias     string
+	ClusterID string
+	Port      string
+	InUsePort string
+	Upstream  []string
+}
+
+func setTransportDialer(rec *Reconciler) {
```

Comment:
dumb question, but why this is not `func (rec *Reconciler) setTransportDialer()`? it seems that is actually builds a transport, so it might be `func buildTransport(rec *Reconciler) http.RountTripper` ?

Reply:
Reworked the whole object creation process.